### PR TITLE
west: Add a board-reset command

### PR DESCRIPTION
**Pull request by galak**
Add support for 'west reset' which will reset a board if supported by
the runner.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

File: scripts/west_commands/reset.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class Reset(WestCommand):
+
+    def __init__(self):
+        super(Reset, self).__init__(
+            'reset',
+            # Keep this in sync with the string in west-commands.yml.
+            'reset and run a binary on a board',
+            "Permanently reprogram a board's reset with a new binary.",
```
Comment (carlescufi) [inline scripts/west_commands/reset.py:21]: This text is wrong isn't it? This doesn't reprogram?

Comment (galak) [reply scripts/west_commands/reset.py:21]: Copied from the flash command, will fix that.

File: boards/common/pyocd.board.cmake [hunk 2]
```diff
@@ -2,4 +2,5 @@
 
 board_set_flasher_ifnset(pyocd)
 board_set_debugger_ifnset(pyocd)
+board_set_reset_ifnset(pyocd)
```
Comment (mbolivar-nordic) [inline boards/common/pyocd.board.cmake:5]: For consistency with the other `board_set_<foo>er_ifnset` functions, I think this should be named `board_set_resetter_ifnset()`.

File: cmake/extensions.cmake [hunk 3]
```diff
@@ -644,6 +644,11 @@ macro(board_set_debugger runner)
   board_set_runner(DEBUG ${runner})
 endmacro()
 
+# A convenience macro for board_set_runner(RESET ${runner}).
+macro(board_set_debugger runner)
```
Comment (mbolivar-nordic) [inline cmake/extensions.cmake:648]: copy/paste bug

**Review by mbolivar-nordic — COMMENTED** [review]
Looks sane to me. Just a couple of things.

**Comment by mbolivar-nordic** [comment]
Hmm... actually, I don't know about the name `reset`. That feels like the sort of name we ought to reserve for west and not use as an extension, since there is a git command by the same name.

How about `west reset-board` or something like that?

**Comment by galak** [comment]
@mbolivar-nordic let me know if you think we need to rename the cmake side of things.

File: cmake/flash/CMakeLists.txt [hunk 4]
```diff
@@ -133,7 +137,7 @@ endif()
 
 # Generate the flash, debug, debugserver, attach targets within the build
 # system itself.
-foreach(target flash debug debugserver attach)
+foreach(target flash debug debugserver attach reset)
```
Comment (tejlmand) [inline cmake/flash/CMakeLists.txt:140]: Either remove, or as minimum have same name as the `west` command.

**Review by tejlmand — CHANGES_REQUESTED** [review]
> @mbolivar-nordic let me know if you think we need to rename the cmake side of things.

I would like the CMake target and west command to have same name.
Anything else is bound to give confusion.

Now that we have `west` extension commands, then the question is, should we still copy those commands into the CMake build system ?
I would actually prefer to teach users to use west directly, especially as `west` allows to provide additional arguments, which `ninja` doesn't.

The only reason for keeping `flash`, `debug`, etc. is for historical reasons.
See also this thread: https://github.com/zephyrproject-rtos/zephyr/pull/28047#discussion_r486238577
where a user will need to provide `--erase`, like `west flash --erase` which is not possible to do with `ninja flash`

So my preference is actually to only provide this functionality as a `west` command, and not as build target.

**Comment by mbolivar-nordic** [comment]
> So my preference is actually to only provide this functionality as a `west` command, and not as build target.

That seems fine to me @tejlmand

**Review by mbolivar-nordic — COMMENTED** [review]
Looks good except for a couple of things:

- no need for a build system target IMO either
- west-completion.bash needs update

**Pull request closed without merging**